File: scrape/detailed.py
"""Fetch detailed information about products."""

import os
from typing import List
import concurrent.futures

import pymongo
import requests
from pymongo.mongo_client import MongoClient

from proxy import Proxy


_CLIENT = MongoClient(
    f'mongodb+srv://{os.environ.get("mongodb_username")}:{os.environ.get("mongodb_password")}'
    f'@vinskraper.wykjrgz.mongodb.net/'
    f'?retryWrites=true&w=majority&appName=vinskraper'
)
_DATABASE = _CLIENT['vinskraper']['vin']

_PROXY = Proxy()

_DETAILS = 'https://www.vinmonopolet.no/vmpws/v3/vmp/products/{}?fields=FULL'


def _product(product: int) -> dict:
    """Extract detailed information about a single product."""
    for _ in range(10):
        try:
            details = requests.get(_DETAILS.format(product),
                                   proxies=_PROXY.get(),
                                   timeout=3)
            if details.status_code != 200:
                raise ValueError()

            details = details.json()
            return {
                'index': int(details.get('code', 0)),

                'farge': details.get('color', '-'),
                'karakteristikk': [characteristic['readableValue'] for characteristic in
                                   details.get('content', {}).get('characteristics', [])],
                'ingredienser': [ingredient['readableValue'] for ingredient in
                                 details.get('content', {}).get('ingredients', [])],
                'egenskaper': [f'{trait["name"]}: {trait["readableValue"]}' for trait in
                               details.get('content', {}).get('traits', [])],
                'lukt': details.get('smell', '-'),
                'smak': details.get('taste', '-'),

                'passer til': [element['name'] for element in details.get('isGoodFor', [])],
                'lagring': details.get('storagePotential', {}).get('formattedValue', '-'),
                'kork': details.get('cork', '-'),

                'beskrivelse': {'lang': details.get('style', {}).get('description', '-'),
                                'kort': details.get('style', {}).get('name', '-')},
                'metode': details.get('method', '-'),
                'årgang': details.get('year', '-'),
            }
        except Exception as err:
            print(f'{product}: Trying another proxy. {err}')
            _PROXY.renew()

    raise ValueError('Failed to fetch product information.')


def detailed(products: List[int] = None, max_workers=5):
    """
    Fetch detailed information about products and store the results to the database.

    Parameters
    ----------
    products : List[int]
        The products to fetch detailed information about.
        If None, all products are fetched.
    max_workers : int
        The maximum number of (parallel) workers to use when fetching products.

    Raises
    ------
    ValueError
        If the product information could not be fetched.

    Notes
    -----
    The function fetches detailed information about the products in the list `products`.
    If `products` is None, all products are fetched.
    The function uses a thread pool to fetch the products in parallel.
    To prevent memory issues, the function fetches the products in batches.
    """
    if not products:
        print('Fetching all products.')
        products = _DATABASE.distinct('index')

    step = max(len(products) // 1000, 500)
    for i in range(0, len(products), step):
        print(f'Processing products {i} to {i + step}.')

        operations = []
        with concurrent.futures.ThreadPoolExecutor(max_workers=max_workers) as executor:
            results = {executor.submit(_product, product): product for product in products[i:i + step]}
            for future in concurrent.futures.as_completed(results):
                product = future.result()
                if not product:
                    continue
                operations.append(
                    pymongo.UpdateOne(
                        {'index': product['index']},
                        {'$set': product},
                        upsert=True
                    )
                )
        _DATABASE.bulk_write(operations)


if __name__ == '__main__':
    detailed()

File: scrape/detailed.mjs
import axios from "axios";
import { HttpsProxyAgent } from "https-proxy-agent";
import { MongoClient, ServerApiVersion } from "mongodb";
import dotenv from "dotenv";

dotenv.config();

const client = new MongoClient(
  `mongodb+srv://${process.env.MONGO_USR}:${process.env.MONGO_PWD}@snublejuice.faktu.mongodb.net/?retryWrites=true&w=majority&appName=snublejuice`,
  {
    serverApi: {
      version: ServerApiVersion.v1,
      strict: true,
      deprecationErrors: true,
    },
  },
);
await client.connect();

const database = client.db("snublejuice");
const itemCollection = database.collection("products");

const proxies = process.env.PROXY_IPS.split(",").flatMap((ip) => [
  new HttpsProxyAgent(
    `http://${process.env.PROXY_USR}:${process.env.PROXY_PWD}@${ip}:${process.env.PROXY_PRT}`,
  ),
]);

function getNextProxy(proxy) {
  if (proxies.length === 0) {
    throw new Error("No more proxies available!");
  }

  const index = proxy ? proxies.findIndex((p) => p.proxy.host === proxy.proxy.host) : -1;
  return proxies[(index + 1) % proxies.length];
}

function removeProxy(proxy) {
  const index = proxies.findIndex((p) => p.proxy.host === proxy.proxy.host);
  if (index > -1) {
    const removedProxy = proxies.splice(index, 1)[0];
    console.log(
      `Removed failing proxy ${removedProxy.proxy.host}. ${proxies.length} proxies remaining.`,
    );
  }
}

const URL = "https://www.vinmonopolet.no/vmpws/v3/vmp/products/{}?fields=FULL";

function processInformation(product) {
  const processed = {
    index: parseInt(product.code, 10) || null,

    updated: true,

    volume: product.volume?.value || 0.0,
    price: product.price?.value || 0.0,

    colour: product.color || null,

    characteristics:
      product.content?.characteristics?.map((characteristic) => characteristic.readableValue) || [],
    ingredients: product.content?.ingredients?.map((ingredient) => ingredient.readableValue) || [],
    ...product.content?.traits?.reduce(
      (acc, trait) => ({ ...acc, [trait.name.toLowerCase()]: trait.readableValue }),
      {},
    ),
    smell: product.smell || null,
    taste: product.taste || null,
    allergens: product.allergens || null,

    pair: product.content?.isGoodFor?.map((element) => element.name) || [],
    storage: product.content?.storagePotential?.formattedValue || null,
    cork: product.cork || null,

    alcohol: product.traits?.find((trait) => trait.name === "Alkohol")?.readableValue || null,
    sugar: product.traits?.find((trait) => trait.name === "Sukker")?.readableValue || null,
    acid: product.traits?.find((trait) => trait.name === "Syre")?.readableValue || null,

    description: {
      lang: product.content?.style?.description || null,
      short: product.content?.style?.name || null,
    },
    method: product.method || null,
    year: product.year || null,
  };

  if (processed.volume > 0 && processed.price > 0) {
    processed.literprice = (processed.price / processed.volume) * 100;
  } else {
    processed.literprice = null;
  }

  // Check if "alkohol" or "alcohol" is present in the processed object.
  if (processed.alkohol || processed.alcohol) {
    // Split the string at the first space character, and convert to float.
    if (processed.alkohol) {
      processed.alcohol = parseFloat(processed.alkohol.split(" ")[0].replace(",", "."));

      // Remove the "alkohol" key from the object.
      delete processed.alkohol;
    } else {
      processed.alcohol = parseFloat(processed.alcohol.split(" ")[0].replace(",", "."));
    }

    // Calculate the alcohol price.
    processed.alcoholprice = processed.literprice / processed.alcohol;
  } else {
    processed.alcohol = 0.0;
    processed.alcoholprice = null;
  }

  return processed;
}

async function getInformation(id, proxy) {
  for (let attempt = 0; attempt < 5; attempt++) {
    try {
      const response = await session.get(URL.replace("{}", id), {
        // httpsAgent: proxy,
        timeout: 10000,
      });

      if (response.status === 200) {
        return processInformation(response.data);
      }

      console.log(`Status code ${response.status} for id ${id} (trying another proxy)`);
    } catch (err) {
      console.log(`Request failed with proxy ${proxy.host} for id ${id}: ${err.message}`);
      if (err.code === "ECONNREFUSED" || err.code === "ETIMEDOUT") {
        removeProxy(proxy);
      }
    }

    try {
      proxy = getNextProxy(proxy);
      await new Promise((resolve) => setTimeout(resolve, 2000));
    } catch (err) {
      throw new Error(`No more proxies available to fetch id ${id}`);
    }
  }

  throw new Error(`Failed to fetch id ${id} after 5 attempts.`);
}

async function updateDatabase(data) {
  const operations = data.map((record) => ({
    updateOne: {
      filter: { index: record.index },
      update: { $set: record },
      upsert: true,
    },
  }));

  return await itemCollection.bulkWrite(operations);
}

async function updateInformation(itemIds) {
  let items = [];
  let proxy = getNextProxy();

  console.log(`Fetching ${itemIds.length} items.`);
  for (const element of itemIds) {
    const id = element["index"];
    console.log(`Id ${id}.`);

    try {
      let product = await getInformation(id, proxy);
      if (!product) {
        console.log(`Unable to find product of Id ${id}. Aborting.`);
        break;
      }

      items.push(product);
      proxy = getNextProxy(proxy);

      await new Promise((resolve) => setTimeout(resolve, 1100));
    } catch (err) {
      console.log(`Error! ${err}`);
      break;
    }

    // Upsert to the database every 10 items.
    if (items.length >= 10) {
      console.log(`Adding ${items.length} products.`);
      const result = await updateDatabase(items);
      console.log(` Modified ${result.modifiedCount} records`);
      console.log(` Upserted ${result.upsertedCount} records`);

      items = [];
    }
  }

  // Insert the remaining products, if any.
  if (items.length === 0) {
    return;
  }
  console.log(`Adding ${items.length} final products.`);
  const result = await updateDatabase(items);
  console.log(` Modified ${result.modifiedCount} records`);
  console.log(` Upserted ${result.upsertedCount} records`);
}

const session = axios.create();

async function main() {
  const itemIds = await itemCollection
    .find({ description: null })
    .project({ index: 1, _id: 0 })
    .toArray();

  await updateInformation(itemIds);
}

await main();

client.close();
